File: src/main/java/de/sean/blockprot/bukkit/listeners/HopperEventListener.java
package de.sean.blockprot.bukkit.listeners;

import de.sean.blockprot.bukkit.BlockProt;
import de.sean.blockprot.bukkit.nbt.BlockNBTHandler;
import org.bukkit.block.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryMoveItemEvent;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.inventory.InventoryHolder;
import org.jetbrains.annotations.Nullable;

public class HopperEventListener implements Listener {
    @EventHandler
    public void onItemMove(InventoryMoveItemEvent event) {
        if (event.getSource().getHolder() == null) return;
        if (BlockProt.getDefaultConfig().isWorldExcluded(event.getSource().getHolder())) return;
        if (event.getDestination().getType() == InventoryType.HOPPER) {
            // This is a hopper trying to pull from something.
            if (BlockProt.getDefaultConfig().isLockableInventory(event.getSource().getType())) {
                Block source = getBlock(event.getSource().getHolder());
                if (source != null) {
                    if (!(new BlockNBTHandler(source)).getRedstone()) {
                        event.setCancelled(true);
                    }
                }
            }
        }
    }

    /**
     * Get the block from a InventoryHolder, instead of just getting the block by location as before.
     */
    @Nullable
    private Block getBlock(InventoryHolder holder) {
        if (holder instanceof Chest) return ((Chest) holder).getBlock();
        else if (holder instanceof Furnace) return ((Furnace) holder).getBlock();
        else if (holder instanceof Hopper) return ((Hopper) holder).getBlock();
        else if (holder instanceof Barrel) return ((Barrel) holder).getBlock();
        else if (holder instanceof BrewingStand) return ((BrewingStand) holder).getBlock();
        else if (holder instanceof ShulkerBox) return ((ShulkerBox) holder).getBlock();
        else if (holder instanceof DoubleChest) {
            DoubleChest doubleChest = (DoubleChest) holder;
            if (doubleChest.getWorld() == null) return null;
            return doubleChest.getWorld().getBlockAt(doubleChest.getLocation());
        } else return null;
    }
}

File: src/main/kotlin/de/sean/blockprot/bukkit/listeners/HopperEventListener.kt
/*
 * Copyright (C) 2021 spnda
 * This file is part of BlockProt <https://github.com/spnda/BlockProt>.
 *
 * BlockProt is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BlockProt is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with BlockProt.  If not, see <http://www.gnu.org/licenses/>.
 */
package de.sean.blockprot.bukkit.listeners

import de.sean.blockprot.bukkit.BlockProt
import de.sean.blockprot.bukkit.nbt.BlockNBTHandler
import org.bukkit.block.*
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryMoveItemEvent
import org.bukkit.event.inventory.InventoryType
import org.bukkit.inventory.BlockInventoryHolder
import org.bukkit.inventory.InventoryHolder

class HopperEventListener : Listener {
    @EventHandler
    fun onItemMove(event: InventoryMoveItemEvent) {
        if (BlockProt.getDefaultConfig().isWorldExcluded((event.source.holder as BlockInventoryHolder).block.world)) return
        if (event.destination.type == InventoryType.HOPPER) {
            // This is a hopper trying to pull from something.
            when {
                BlockProt.getDefaultConfig().isLockableInventory(event.source.type) -> {
                    // This hopper is trying to pull from some inventory which *may* be locked.
                    // Note: we do not have to check for double chests, as both sides of a chest are individually locked.
                    val sourceLocation = getBlock(event.source.holder ?: return) ?: return
                    val handler = BlockNBTHandler(sourceLocation)
                    if (!handler.redstone) event.isCancelled = true
                }
                else -> return
            }
        }
    }

    /**
     * Get the block from a InventoryHolder, instead of just getting the block by location as before.
     */
    private fun getBlock(holder: InventoryHolder): Block? {
        return when (holder) {
            is Chest -> holder.block
            is Furnace -> holder.block
            is Smoker -> holder.block
            is BlastFurnace -> holder.block
            is Hopper -> holder.block
            is Barrel -> holder.block
            is BrewingStand -> holder.block
            is ShulkerBox -> holder.block
            is DoubleChest -> holder.world?.getBlockAt(holder.location)
            else -> null
        }
    }
}
